File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(maps):
    q = deque([(0, 0, 1)])
    n = len(maps)
    m = len(maps[0])
    visited = [[0 for _ in range(m)] for _ in range(n)]
    visited[0][0] = 1
    
    while q:
        x, y, dist = q.popleft()
        if x == n-1 and y == m-1:
            return dist
        
        for d in range(4):
            nx = x + dx[d]
            ny = y + dy[d]
            
            if 0<=nx<n and 0<=ny<m:
                if not visited[nx][ny] and maps[nx][ny] == 1:
                    q.append((nx, ny, dist+1))
                    visited[nx][ny] = 1
    return False
            

def solution(maps):
    answer = bfs(maps)
    
    if answer:
        return answer
    return -1
    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    
    int[] dx = {-1, 1, 0, 0};
    int[] dy = {0, 0, -1, 1};
    
    int bfs(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        
        boolean[][] visited = new boolean[n][m];
        ArrayDeque<int[]> q = new ArrayDeque<>();
        q.add(new int[]{0, 0, 1});
        visited[0][0] = true;
        int[] status = null;
        boolean flag = false;
        while(!q.isEmpty()) {
            status = q.poll();
            if(status[0] == n-1 && status[1] == m-1) {
                return status[2];
            }
            for(int i=0; i<4; i++) {
                int nx = status[0] + dx[i];
                int ny = status[1] + dy[i];
                
                if(0<=nx && nx<n && 0<=ny && ny < m){
                    if(maps[nx][ny] == 1 && visited[nx][ny] == false) {
                        visited[nx][ny] = true;
                        q.add(new int[]{nx, ny, status[2]+1});
                    }
                }
            }
        }
        return -1;
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        return bfs(maps);
    }
}